File: java-simple-http-server/src/main/java/MimeDetector.java
import java.io.IOException;
import java.util.Properties;

public class MimeDetector {

    private final Properties prop;

    public MimeDetector(String configFileName) {
        var in = this.getClass().getResourceAsStream(configFileName);
        var props = new Properties();
        try {
            props.load(in);
        } catch (IOException e) {
            System.err.println("Failed to load mime config");
        }
        this.prop = props;
    }

    /**
     * 与えられたファイル名の拡張子に対応するMIMEを返す。
     * 該当する拡張子が無い場合はapplication/octet-streamを返す。
     */
    public String getMime(String fileName) {
        var ext = fileName.substring(fileName.indexOf(".") + 1);
        return this.prop.getProperty(ext, "application/octet-stream");
    }
}

File: scala-simple-http-server/src/main/scala/MimeDetector.scala
import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Path, Paths}

import scala.collection.JavaConverters._
import scala.io.Source
import scala.util.matching.Regex
import scala.util.parsing.combinator._

class MimeDetector(configFileName: String) {

  private val mimeMap = MimeConfParser.parse(Source.fromResource(configFileName).mkString)
  private val extPattern = """(?<=\.)[A-Za-z0-9]+$""".r

  def getMime(fileName: String): String =
    extPattern.findFirstIn(fileName) // Regex#findFirstInはOptionを返す
      .flatMap(ext => mimeMap.get(ext)) // Map#getもOptionを返す。なのでflatMapする
      .getOrElse("application/octet-stream")
}

object MimeConfParser extends RegexParsers {

  override val whiteSpace: Regex = """(\s|#.*)+""".r

  // values
  private def key = """[\w\./+-]+""".r
  private def value = repsep("""[\w\./+-]+""".r, """\s?""".r)
  private def line = key ~ value <~ ";"
  private def list = """types\s*\{""".r ~> rep(line) <~ "}"

  def parse(content: String): Map[String, String] = {
    parseAll(list, content) match {
      case Success(result, _) => result.flatMap { case MimeConfParser.~(mime, exts) => exts.map(_ -> mime) }.toMap
      case NoSuccess(msg, next) => throw new ConfigParseException(s"Invalid config: $msg on line ${next.pos.line} on column ${next.pos.column}")
    }
  }

  class ConfigParseException(msg: String) extends RuntimeException(msg)
}
